src/amber/router/params.cr

```diff
@@ -1,20 +1,104 @@
+require "http"
+require "./parsers/*"
+require "./file"
+
 module Amber::Router
-  class Params < Hash(String, String)
-    alias KeyType = String | Symbol
-    property files = {} of String => Amber::Router::Files::File
+  module Types
+    alias Key = String | Symbol
+    alias Files = Hash(String, Amber::Router::File)
+    alias Params = Hash(String, String)
+  end
+
+  class Params
+    TYPE_EXT_REGEX   = Amber::Support::MimeTypes::TYPE_EXT_REGEX
+    URL_ENCODED_FORM = "application/x-www-form-urlencoded"
+    MULTIPART_FORM   = "multipart/form-data"
+    APPLICATION_JSON = "application/json"
+
+    getter empty_params = Types::Params.new
+
+    @files = Types::Files.new
+    @multipart : Types::Params?
+    @json : Types::Params?
+    @form : HTTP::Params?
+
+    def initialize(@request : HTTP::Request)
+    end
+
+    def [](key : Types::Key) : String
+      self.[key]? || raise Amber::Exceptions::Validator::InvalidParam.new(key)
+    end
+
+    def []?(key : Types::Key)
+      _key = key.to_s
+      route[_key]? || override_method?(_key) || json[_key]?
+    end
+
+    def files
+      multipart
```

Comment:
Check `if @multipart` before calling this is faster and more readable than doing the check in the method.